sacc2/sacc.py

```diff
@@ -0,0 +1,871 @@
+
+        **tags:
+            Tags to apply to this data point.
+            Tags can be any arbitrary metadata that you might want later,
+            For 2pt data the tag would include an angle theta or ell.
+
+        Returns
+        -------
+        None
+        """
+        if self.covariance is not None:
+            raise ValueError("You cannot add a data point after setting the covariance")
+        tracers = tuple(tracers)
+        for tracer in tracers:
+            if (tracer not in self.tracers) and (not tracers_later):
+                raise ValueError(f"Tracer named '{tracer}' is not in the known list of tracers."
+                                 "Either put it in before adding data points or set tracers_later=True")
+        d = DataPoint(data_type, tracers, value, **tags)
+        self.data.append(d)
+
+    def add_covariance(self, covariance):
+        """
+        Once you have finished adding data points, add a covariance
+        for the entire set.
+
+        Parameters
+        ----------
+        covariance: array or list
+            2x2 numpy array containing the covariance of the added data points
+            OR a list of blocks
+
+        Returns
+        -------
+        None
+        """
+        if isinstance(covariance, BaseCovariance):
+            self.covariance = covariance
+        else:
+            self.covariance = BaseCovariance.make(covariance, len(self))
+
+    def _indices_to_bool(self, mask):
+        # Convert an array of indices into a boolean True mask
+        if mask.dtype not in [np.int8, np.int16, np.int32, np.int64]:
+            raise ValueError(f"Wrong mask type ({mask.  dtype}) - expected integers or boolean")
+        m = np.zeros(len(self), dtype=bool)
+        for i in mask:
+            m[i] = True
+        return m
+
+    def mask(self, mask):
+        """
+        Select data points, keeping only values where the mask is True or an index is
+        included in it.
+
+        You can use Sacc.cut to do the opposite operation, keeping points where the mask
+        is False.
+
+        You can find indices (e.g. matching some tag) to keep using the Sacc.indices method.
+
+        Parameters
+        ----------
+        mask: array or list
+            Mask must be either a boolean array or a list/array of integer indices to remove.
+            if boolean then True means to keep a data point and False means to cut it
+            if integers then values indicate data points to keep
+        """
+        mask = np.array(mask)
+
+        # Convert integer masks to booleans
+        if mask.dtype != np.bool:
+            mask = self._indices_to_bool(mask)
+
+        self.data = [d for i, d in enumerate(self.data) if mask[i]]
+        if self.covariance is not None:
+            self.covariance = self.covariance.masked(mask)
+
+    def cut(self, cut):
+        """
+        Remove data points, getting rid of points where the mask is True or an index is
```

Comment:
Yep, exactly - I was thinking of calls to `indices` followed by `cut` or mask.
Definitely a good idea to have something that does both together, since it's a common pattern - maybe rename `cut` to `cut_indices` and have a `cut` method that does both?

Reply:
Sounds good! Having specific "scale_cut" functions where you just pass an array of ell_min and ell_max for each C_ell (and thetas for correlation functions) would also be useful. And a similar one for "cut all points involving tracer X".

Reply:
I've done the first part of this, renaming the functions and adding.  I'd like to punt a little on the helper functions in the interests of getting through this, though they're definitely handy.

Reply:
ok, sounds good!